asm-multiproject/istio-multiproject/istio-operator.yaml

```diff
@@ -0,0 +1,51 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: install.istio.io/v1alpha1
+kind: IstioOperator
+metadata:
+  annotations:
+    gke.io/cluster: "gke://PROJECT_ID/us-central1-c/asm-cluster" # {"$ref":"#/definitions/io.k8s.cli.substitutions.gkeio-cluster"}
+  name: default-istio-operator
+spec:
+  profile: asm-gcp-multiproject # {"$ref":"#/definitions/io.k8s.cli.setters.anthos.servicemesh.profile"}
+  hub: gcr.io/gke-release/asm # {"$ref":"#/definitions/io.k8s.cli.setters.anthos.servicemesh.hub"}
+  tag: 1.6.5-asm.7 # {"$ref":"#/definitions/io.k8s.cli.setters.anthos.servicemesh.tag"}
+  components:
+    pilot:
+      k8s:
+        env:
+        - name: PILOT_SKIP_VALIDATE_TRUST_DOMAIN
+          value: "true"
+        - name: SPIFFE_BUNDLE_ENDPOINTS
+          value: "PROJECT_ID.svc.id.goog|https://storage.googleapis.com/mesh-ca-resources/spiffe_bundle.json" # {"$ref":"#/definitions/io.k8s.cli.substitutions.spiffe-bundle-endpoints"}
+  meshConfig:
+    defaultConfig:
+      proxyMetadata:
+        GCP_METADATA: "PROJECT_ID|PROJECT_NUMBER|asm-cluster|us-central1-c" # {"$ref":"#/definitions/io.k8s.cli.substitutions.gke-metadata"}
+        TRUST_DOMAIN: "PROJECT_ID.svc.id.goog" # {"$ref":"#/definitions/io.k8s.cli.substitutions.trust-domain"}
+        GKE_CLUSTER_URL: "https://container.googleapis.com/v1/projects/PROJECT_ID/locations/us-central1-c/clusters/asm-cluster" # {"$ref":"#/definitions/io.k8s.cli.substitutions.gke-cluster-url"}
+    trustDomainAliases: # {"$ref":"#/definitions/io.k8s.cli.setters.anthos.servicemesh.trustDomainAliases"}
```

Comment:
I'm still not sure why we cannot reuse the asm directory and only change the asm/cluster/istio-operator.yaml?

It looks like we can just have the default empty values for trustDomainAliases and clusterName, or maybe there are some other changes we can not combine there?

Reply:
This is definitely a possibility. Actually I asked Jianfei about merging this two profile together. The issue here is we need to put something under trustDomainAlias (we cannot assign an empty array to the setter). If we think it is safe to put some stuff under trustDomainAlias (I was thinking to put "PROJECT_ID.svc.id.goog", but it looks like it will need some tests to confirm it will work), then we probably can merge them together.

Reply:
it can just be an empty string by default, @incfly  to confirm. And if that's the only change plus a few more in istio-operator.yaml to support the multi-project profile, we should try to add them to asm, having 8 duplicate directories with only minor changes in istio-operator.yaml is just crazy...

Reply:
An empty array for trust domain alias should be safe (for non-meshca-multi-project cases).

But @yanyuan-meiyi have you mentioned there're some kpt issues about that? I don't recall what it is, sometihng like set it with some value, unable to set it back to empty or something. I'm not sure.

Reply:
(my first reply is too fast... deleted)

I verified that the alias as same as the trust domain would be a no-op, compared with not setting the aliases at all. That should address Yuan's concern of not able to resetting it back to empty.

```
trustDomain: "proj.id.svc.goog"
trustDomainAliases: ["proj.id.svc.goog"]
```

Reply:
> (my first reply is too fast... deleted)
> 
> I verified that the alias as same as the trust domain would be a no-op, compared with not setting the aliases at all. That should address Yuan's concern of not able to resetting it back to empty.
> 
> ```
> trustDomain: "proj.id.svc.goog"
> trustDomainAliases: ["proj.id.svc.goog"]
> ```

Thanks! So we will then merge asm-gcp and asm-gcp-multiproject into one kpt package. Other than the network configurations, we will add two more setters: trustdomainalias and an additional environment variable PILOT_SKIP_VALIDATE_TRUST_DOMAIN. By default (asm-gcp), trustdomainalias will be the same as trustdomain, PILOT_SKIP_VALIDATE_TRUST_DOMAIN will be set as "false", so users using asm-gcp don't need to worry about these new settings. @elfinhe, does it sound good to you?

Reply:
Thanks! So we will then merge asm-gcp and asm-gcp-multiproject into one kpt package. Other than the network configurations, we will add two more setters: trustdomainalias and an additional environment variable PILOT_SKIP_VALIDATE_TRUST_DOMAIN. By default (asm-gcp), trustdomainalias will be the same as trustdomain, PILOT_SKIP_VALIDATE_TRUST_DOMAIN will be set as "false", so users using asm-gcp don't need to worry about these new settings. @elfinhe, does it sound good to you?

Reply:
Sounds good if the behaviors are the same with `trustDomainAliases: [<trust-domain>]` and  `trustDomainAliases:`.

Reply:
Actually, my previous comment was wrong, I think the user needs to configure trustdomainalias if we merge the packages together in this way. Right now, we use substitution in kpt to populate trust domain based on this pattern `PROJECT_ID.id.svc.goog`. However, we don't have something like substitution array. We need to use setters. That means the customer needs to manually set this field.

Reply:
hmm, can we work around by writing yamls as 

```yaml
trustDomainAliases:
- PROJECT_ID.svc.goog.id
```

If the issue is about kpt's setters in the string prefix in a array. (on the other hand, i'm wondering what's the standard kpt's answer for this kinda flexiblity, we have run into multiple syntax limitations so far)

Reply:
Hmm.. For setters, I don't think we can do that. The setter will overwrite the whole string. Let me ask if there is something can behave like an array of substitution.

Reply:
Just pinged Phani on this. Right now, we cannot do something like "array of substitution" using kpt.